virtwho/pid_lock.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+from __future__ import print_function, absolute_import
+"""
+Agent for reporting virtual guest IDs to subscription-manager
+
+Copyright (C) 2011 Radek Novacek <rnovacek@redhat.com>
+
+This program is free software; you can redistribute it and/or
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; either version 2
+of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+"""
+
+import sys
+import os
+import time
+
+PIDFILE = "/var/run/virt-who.pid"
+STATUS_LOCK = "/var/run/virt-who-status.pid"
+STATUS_DATA = "/var/lib/virt-who/run_data.json"
+STATUS_DATA_DIR = "/var/lib/virt-who"
+
+
+class PIDLock(object):
+    def __init__(self, filename, wait_to_lock=False):
+        self.filename = filename
+        self.wait_to_lock = wait_to_lock
+
+    def is_locked(self):
+        try:
+            with open(self.filename, "r") as f:
+                pid = int(f.read().strip())
+            try:
+                os.kill(pid, 0)
+                return True
+            except OSError:
+                # Process no longer exists
+                print("PID file exists but associated process " \
+                                     "does not, deleting PID file", file=sys.stderr)
+                os.remove(self.filename)
+                return False
+        except Exception:
+            return False
+
+    def __enter__(self):
+        # ensure file not locked
+        if self.wait_to_lock:
+            while self.is_locked():
+                time.sleep(1)
```

Comment:
It seems that `PIDLock` was implemented to avoid running multiple instances of virt-who. 

I think that this approach has danger of race conditions, because there is no atomicity of check (file exists) and set (save current PID to file). There is no timeout in while loop and one process can be stuck forever. I'm not sure about sleeping for one second.

I would rather use approach described here:

https://stackoverflow.com/questions/489861/locking-a-file-in-python

Example:

```python
from filelock import FileLock
import os
import time


def main():
    pid = os.getpid()
    print(f'{pid} is waiting for lock')
    with FileLock("myfile.txt.lock"):
        print(f"{pid} lock acquired...")
        with open("myfile.txt", "w+") as my_file:
            my_file.write(str(pid))
            time.sleep(2.0)
    print(f"{pid} lock released")


if __name__ == '__main__':
    main()
```

and then try to run it using: `python ./example.py & python ./example.py`

Reply:
Oh, this is not part of Python. It requires another RPM :-( and the implementation is not too much different from the implementation in `pid_lock.py`.